staging/vhost-device-gpu/src/main.rs

```diff
@@ -56,13 +56,15 @@ impl TryFrom<GpuArgs> for GpuConfig {
 
 fn start_backend(config: GpuConfig) -> Result<()> {
 
-    let handle: JoinHandle<Result<()>> = spawn(move || loop {
+    let handle: JoinHandle<Result<()>> = spawn(move || {
+        let virtio_gpu = VirtioGpu::new();
```

Comment:
Have you checked that handle_event is being called from the thread in which we create the `virtio_gpu` here?
I briefly looked at the vhost_backend code and it seemed like it might be called from a different thread.

Reply:
at this point the thread id doesn't really matter because we are passing that virtio_gpu instance to VhostUserGpuBackend::new() which is later referenced as self.virtio_gpu

Reply:
This code spawns a thread that executes this block once, the previous design with the loop continues until the thread is explicitly terminated. if you put this code in the handle_event when handle_event is being triggered again it will error already in use which is why I impl send and sync for rutabaga so that we can share this.

Reply:
> at this point the thread id doesn't really matter because we are passing that virtio_gpu instance to VhostUserGpuBackend::new() which is later referenced as self.virtio_gpu

I disagree, as far as I understand rutabaga the thread matters a lot.

If the rutabaga context is not supposed to be Send+Sync (which I think the authors have done on purpose), calling it from a different is likely to be undefined behavior.

Reply:
> This code spawns a thread that executes this block once, the previous design with the loop continues until the thread is explicitly terminated. if you put this code in the handle_event when handle_event is being triggered again it will error already in use which is why I impl send and sync for rutabaga so that we can share this.

Yes I know why you impl Send+Sync, but I think the implementation is unsound*. And if handle_event is called from a different thread and calls to rutabaga from this different thread, it is likely undefined behavior.


* By unsound I mean that the unsafe impl might cause safe code to cause undefined behavior.

Reply:
I don't think it will require the VirtioGpu to be Send+Sync, but that doesn't matter because:

"from what i can see handle event is not being called from thesame thread as in main. we will need to figure out other options otherwise I suggest to keep this workaround impl send+sync for now."

This doesn't make sense to me, if you do unsafe impl send+sync, and it is being called from a different thread, then yes it will compile, but that doesn't mean it is correct, it would basically just cause undefined behavior and solve nothing, maybe some assertion fails somewhere, maybe some runtime crash or some memory corruption. Doing `unsafe impl Send` and `unsafe impl Sync` basically tells the compiler to "trust me you can use it from a different thread", we know you can't use rutabaga from a different thread and (if I understood you corectly) we know handle_event is being called from a different thread.